scipy/sparse/_coo.py

```diff
@@ -556,31 +571,94 @@ def _add_dense(self, other):
         dtype = upcast_char(self.dtype.char, other.dtype.char)
         result = np.array(other, dtype=dtype, copy=True)
         fortran = int(result.flags.f_contiguous)
-        M, N = self._shape_as_2d
-        coo_todense(M, N, self.nnz, self.row, self.col, self.data,
-                    result.ravel('A'), fortran)
+        if self.ndim < 3:
+            M, N = self._shape_as_2d
+            coo_todense(M, N, self.nnz, self.row, self.col, self.data,
+                        result.ravel('A'), fortran)
+        else:
+            coords = np.concatenate(self.coords)
+            coo_todense_nd(self.shape, self.nnz, self.ndim, coords,
+                           self.data, result.ravel('A'), fortran)
         return self._container(result, copy=False)
+    
 
-    def _matmul_vector(self, other):
-        result_shape = self.shape[0] if self.ndim > 1 else 1
-        result = np.zeros(result_shape,
-                          dtype=upcast_char(self.dtype.char, other.dtype.char))
+    def _add_sparse(self, other):
+        if self.ndim < 3:
+            return _data_matrix._add_sparse(self, other)
```

Comment:
`_data_matrix` doesn't define `_add_sparse()`, but its parent class `_spbase` does. And that implementation is converting to CSR format.

Let's do the CSR conversion explicitly here. (and in `_sub_sparse`).

Reply:
I've both added the CSR conversion and changed the class we are calling `_add_sparse` from. Addressed at https://github.com/scipy/scipy/pull/21197/commits/3f7278228a79c9e47eefce3459bd065b8847e731.

Reply:
I think @perimosocordiae intended you to inline the one-line code from `_spbase` for both `_add_sparse` and `_sub_sparse`. As currently changed, the `tocsr()` method would get called twice. Instead use something like this:
```python
        if self.ndim < 3:
            return self.tocsr()._add_sparse(other)
```
This way the reader doesn't even have to look up the method (and for a one line method like this it is easier to read this line than to look through the inheritance.)

Reply:
Modified at 1322cb5179982afb46a1df30d19def66b653d7a7.